Support ability to rename fields and deprecate existing names; does so for S3

jeffalder opened the pull request:
## Description
When renaming properties via customization.config, the existing names can be deprecated. 
On S3 CopyObjectRequest, deprecates `bucket` in favor of `destinationBucket` and `key` in favor of `destinationKey`.

## Motivation and Context
This covers half of the requests in #1313 .

## Testing
Added tests for deprecation
Validated resulting CopyObjectResult class

## Screenshots (if appropriate)
n/a

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License
<!--- The SDK is released under the Apache 2.0 license (http://aws.amazon.com/apache2.0/), so any code you submit will be released under that license -->
<!--- For substantial contributions, we may ask you to sign a Contributor License Agreement (http://en.wikipedia.org/wiki/Contributor_License_Agreement) -->
<!--- Put an `x` in the below box if you confirm that this request can be released under the Apache 2 license -->
- [x] I confirm that this pull request can be released under the Apache 2 license

jeffalder commented:
@millems Here ya go! Have at it. Thanks for your lightning-fast turnaround on feedback, too!

millems commented on codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/deprecatedrenamerequest.java:
```diff
@@ -0,0 +1,264 @@
+        public final Builder originalNameDeprecated(String newName) {
+            this.newName = newName;
+            return this;
+        }
```
Minor Nit: Should probably have bean-based methods for the deprecated field. These are around for people using things like spring-xml, so it's likely not going to break anyone to exclude them, but it'd be nice to have for completeness.

jeffalder replied:
I'll try to have this for you by Monday.

jeffalder replied:
I think this is good now.

jeffalder commented:
I added the bean-style setters. Then I realized that there was a gap in the logic for lists and maps; I updated the checks to make sure that lists and maps can't be deprecated (yet), along with enum types.
